keps/sig-scheduling/20180409-scheduling-framework.md

```diff
@@ -198,49 +197,55 @@ post-filter extension point.
 
 ### Scoring
 
-These plugins are used to rank nodes that have passed the filtering phase. The
-scheduler will call each scoring plugin for each node. There will be a well
-defined range of integers representing the minimum and maximum scores. After the
-[normalize scoring](#normalize-scoring) phase, the scheduler will combine node
-scores from all plugins according to the configured plugin weights.
+These plugins have two phases:
 
-### Normalize scoring
-
-These plugins are used to modify scores before the scheduler computes a final
-ranking of Nodes. A plugin that registers for this extension point will be
-called with the [scoring](#scoring) results from the same plugin. This is called
-once per plugin per scheduling cycle.
+1. The first phase is called "score" which is used to rank nodes that have passed 
+the filtering phase. The scheduler will call `Score` of each scoring plugin for
+each node. There will be a well defined range of integers representing the minimum
+and maximum scores. 
```

Comment:
@bsalamat we don't enforce this range currently, is this supposed to be enforced before or after the normalize (if exists)?

Reply:
> @bsalamat we don't enforce this range currently, is this supposed to be enforced before or after the normalize (if exists)?

if we refactor InterPodAffinity with the scheduling framework and the score-normalize pattern, we can't enforce this in the "score" phase.

We could reword to "the scoring plugin will return a node to score map after the score and normalize score phases, which will be a well-defined range of integers...", and put it at the beginning of the section.

But it would be a little more complicated for the user to understand.

Reply:
We can enforce the range at the time we [apply the weights](https://github.com/kubernetes/kubernetes/pull/80383/files#diff-76b7e314f9618c8bc2ca6442ddd35dc3R379).

Reply:
> We can enforce the range at the time we [apply the weights](https://github.com/kubernetes/kubernetes/pull/80383/files#diff-76b7e314f9618c8bc2ca6442ddd35dc3R379).

Yes, we could do that, but the developers are supposed to decide how to enforce the range in this extension point, ex: shrink by the ratio ( `*(max-min)/max`), etc.

Reply:
Yes, they can do that, but the framework should also either error out when the score returned by a plugin is not within the range, or apply a simple default flooring/ceiling logic.

Reply:
I think it could be better to skip the plugin and log an error when the plugins don't enforce the range which wouldn't cause weird behaviour by applying flooring/ceiling logic or scheduling failure by returns an error. WDYT?

Reply:
skipping the plugin and logging an error sounds good to me.

Reply:
I think a score plugin can return any valid integer, but we should expect all the scores to be in a specific range after "normalize score". Either a score plugin implements "normalize" or the scheduler uses a default normalize. If we decide that the range of scores is 0 to 100, then the default normalize would be `100 * (score - min) / (max - min)`.
If the final score after normalize is not in the range, I agree that we should skip the plugin and return an error.

Reply:
@bsalamat Could you take a look at this https://github.com/kubernetes/kubernetes/pull/80901 which updates the NodeScoresList to a map, and we don't need to update the KEP twice if we reach a consensus here. Thanks!

Reply:
I think the semantics should be as follows: 
1) A Score function should return a value in the range [0, 100] unless a normalize score is provided
2) A NormalizeScore should return a value in the range [0, 100] when provided
3) A NormalizeScore is optional to implement, the framework could provide a default implementation, but it should explicitly be called by the plugin (we shouldn't call it by default).

Reply:
> we shouldn't call it by default

+1

But the scheduler will skip the plugin and log an error when the score is not in the range.

Reply:
If we set the range of scores to [0, 100], we should [multiply the framework's score with 0.1](https://github.com/kubernetes/kubernetes/pull/81015/files#diff-c237cdd9e4cb201118ca380732d7f361R811) or priority functions' score with 10.

We can also keep the current range [0, 10] which don't need the annoying multiply operation. And I don't think [0, 10] and [0, 100] would make a difference here.

Reply:
I don't have a reason to change the range from [0, 10] to [0, 100], unless past experiences indicated that the [0, 10] range is too tight, is that the case @bsalamat?

Reply:
We have seen many instances that our scoring functions are unpredictable. While extending the range will not address this problem, it can be one step towards making the score functions more differentiating. One can imagine that in a large cluster with thousands of nodes, a [0, 10] range is not differentiating enough. Now that we are transitioning to the framework, we can consider expanding the range.

Reply:
> We have seen many instances that our scoring functions are unpredictable. While extending the range will not address this problem, it can be one step towards making the score functions more differentiating. One can imagine that in a large cluster with thousands of nodes, a [0, 10] range is not differentiating enough. Now that we are transitioning to the framework, we can consider expanding the range.

OK, I'll multiply the original score with 10.